lib/charms/vault_k8s/v0/vault_managers.py

```diff
@@ -0,0 +1,348 @@
+    AppRole,
+    SecretsBackend,
+    Token,
+    VaultClient,
+    VaultClientError,
+)
+from charms.vault_k8s.v0.vault_tls import File, VaultTLSManager
+from ops import CharmBase, Model, Relation
+
+# The unique Charmhub library identifier, never change it
+LIBID = "4a8652e06ecb4eb28c5fdbf220d126bb"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+AUTOUNSEAL_POLICY = """path "{mount}/encrypt/{key_name}" {{
+    capabilities = ["update"]
+}}
+
+path "{mount}/decrypt/{key_name}" {{
+    capabilities = ["update"]
+}}
+"""
+
+
+class LogAdapter(logging.LoggerAdapter):
+    """Adapter for the logger to prepend a prefix to all log lines."""
+
+    prefix = "vault_managers"
+
+    def process(self, msg, kwargs):
+        """Decides the format for the prepended text."""
+        return f"[{self.prefix}] {msg}", kwargs
+
+
+logger = LogAdapter(logging.getLogger(__name__), {})
+
+
+@dataclass
+class VaultAutounsealRelationDetails:
+    """Computes the auto-unseal related values for a relation.
+
+    This class is used to compute the static details for a vault-autounseal
+    relation, such as the key name, policy name, and approle name. These values
+    are all based on the relation ID.
+
+    This class provides a central place to manage the naming conventions for
+    the auto-unseal functionality.
+    """
+
+    relation: Relation
+
+    @property
+    def key_name(self) -> str:
+        """Return the key name for the relation."""
+        return str(self.relation.id)
+
+    @property
+    def policy_name(self) -> str:
+        """Return the policy name for the relation."""
+        return f"charm-autounseal-{self.relation.id}"
+
+    @property
+    def approle_name(self) -> str:
+        """Return the approle name for the relation."""
+        return f"charm-autounseal-{self.relation.id}"
+
+
+class VaultAutounsealProviderManager:
+    """Encapsulates the auto-unseal functionality.
+
+    This class provides the business logic for auto-unseal functionality in
+    Vault charms. It is opinionated, and aims to make the interface to enabling
+    and managing the feature as simple as possible. Flexibility is sacrificed
+    for simplicity.
+    """
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        juju_model: Model,
+        client: VaultClient,
+        provides: VaultAutounsealProvides,
+        ca_cert: str,
+        mount_path: str = "charm-autounseal",
+    ):
+        self._charm = charm
+        self._juju_facade = JujuFacade(charm)
+        self._model = juju_model
+        self._client = client
+        self._provides = provides
+        self._mount_path = mount_path
+        self._ca_cert = ca_cert
+        self._port = 8200
+
+        if not self._ca_cert:
+            logger.warning("No CA certificate provided for auto-unseal")
```

Comment:
Is this possible as Vault charms are always deployed with TLS? If we go with the assertion approach, shouldn't we simply assert that the ca certificate was provided.